resources/Microsoft.Windows.Developer/Microsoft.Windows.Developer.psm1

```diff
@@ -800,6 +805,79 @@ class WindowsCapability {
         }
     }
 }
+
+[DSCResource()]
+class AdvancedNetworkSharingSetting {
+    [DscProperty(Key, Mandatory)]
+    [AdvancedNetworkSharingSettingName]$Name
+
+    [DscProperty()]
+    [string[]]$Profiles = @()
+
+    [DscProperty(NotConfigurable)]
+    [string[]]$EnabledProfiles
+
+    # Official group names for the firewall rules
+    hidden [string] $NetworkDiscoveryGroup = '@FirewallAPI.dll,-32752'
+    hidden [string] $FileAndPrinterSharingGroup = '@FirewallAPI.dll,-28502'
+
+    [AdvancedNetworkSharingSetting] Get() {
+        $currentState = [AdvancedNetworkSharingSetting]::new()
+        $currentState.Name = $this.Name
+        $currentState.Profiles = $this.Profiles
+
+        if ($this.Name -eq [AdvancedNetworkSharingSettingName]::NetworkDiscovery) {
+            $group = $this.NetworkDiscoveryGroup
+        } else {
+            $group = $this.FileAndPrinterSharingGroup
+        }
+
+        # A firewall group is enabled if none of its sub-rules are disabled and at least one is enabled.
+        $this.EnabledProfiles = Get-NetFirewallRule -Group $group | Group-Object Profile | ForEach-Object {
+            $enabled = ($_.Group.Enabled | Where-Object { $_ -eq 'true' } | Measure-Object).Count
+            $disabled = ($_.Group.Enabled | Where-Object { $_ -eq 'false' } | Measure-Object).Count
+            [PSCustomObject]@{
+                Profile  = $_.Name
+                Count    = $_.Count
+                Enabled  = $enabled
+                Disabled = $disabled
+            }
+        } | Where-Object { ($_.Enabled -gt 0) -and ($_.Disabled -eq 0) -and ($_.Enabled -eq $_.Count) } | Select-Object -Unique -CaseInsensitive -ExpandProperty Profile
+
+        $currentState.EnabledProfiles = $this.EnabledProfiles
+
+        return $currentState
+    }
+
+    [bool] Test() {
+        $currentState = $this.Get()
+
+        # Compare-object is case insensitive by default and does not take null arguments
+        $difference = Compare-Object -ReferenceObject @( $this.Profiles | Select-Object) -DifferenceObject @( $currentState.EnabledProfiles | Select-Object)
```

Comment:
There is a bit of a corner case here, if a rule group is in a mixed state and we are trying to enable it then the right things happen. If it is in a mixed state and we are trying to disable it then we ignore it because the machine is in the desire state from a Enabled bool perspective. 

I am happy to write logic to handle this, but for now I believe that this is the right optimization for us so we do not suddenly start changing untargeted profiles for the user. We can approach this as a bug fix if needed.